### PR TITLE
Handle expression body in PartialMethodCompletionProvider

### LINKED ISSUE #26388: Partial method completion provider broken
**Version Used**: 
Visual Studio 15.6.6, 15.7 Preview 4.0
**Steps to Reproduce**:

```csharp
partial class C
{
    partial void M();
    partial $$
}
```
The completion list provides `M()`, then press enter.

**Expected Behavior**:
`partial void M() => throw new NotImplementedException();` filled.
**Actual Behavior**:
Visual Studio displays an error.

Stack trace in log:
```
System.NullReferenceException: 
Microsoft.CodeAnalysis.CSharp.Completion.Providers.PartialMethodCompletionProvider.GetTargetCaretPosition(SyntaxNode caretTarget)
Microsoft.CodeAnalysis.Completion.Providers.AbstractMemberInsertingCompletionProvider.&lt;GetChangeAsync&gt;d__7.MoveNext()
--- 
System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders.&lt;GetChangeAsync&gt;d__35.MoveNext()
--- 
Roslyn.Utilities.TaskExtensions.WaitAndGetResult_CanCallOnBackground[T](Task`1 task, CancellationToken cancellationToken)
]Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Commit(CompletionItem item, Model model, Nullable`1 commitChar, ITextSnapshot initialTextSnapshot, Action nextHandler)
Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.CommitOnEnter(Boolean&amp; sendThrough, Boolean&amp; committed)
Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Microsoft.VisualStudio.Commanding.IChainedCommandHandler&lt;Microsoft.VisualStudio.Text.Editor.Commanding.Commands.ReturnKeyCommandArgs&gt;.ExecuteCommand(ReturnKeyCommandArgs args, Action nextHandler, CommandExecutionContext context)
Microsoft.VisualStudio.Commanding.CommandHandlerExtensions.ExecuteCommand[T](ICommandHandler commandHandler, T args, Action nextCommandHandler, CommandExecutionContext executionContext)
Microsoft.CodeAnalysis.Editor.CommandHandlers.AbstractCompletionCommandHandler.ExecuteCommandWorker[TCommandArgs](TCommandArgs args, Action nextHandler, CommandExecutionContext context)
Microsoft.CodeAnalysis.Editor.CommandHandlers.AbstractCompletionCommandHandler.Microsoft.VisualStudio.Commanding.IChainedCommandHandler&lt;Microsoft.VisualStudio.Text.Editor.Commanding.Commands.ReturnKeyCommandArgs&gt;.ExecuteCommand(ReturnKeyCommandArgs args, Action nextHandler, CommandExecutionContext context)
Microsoft.VisualStudio.Commanding.CommandHandlerExtensions.ExecuteCommand[T](ICommandHandler commandHandler, T args, Action nextCommandHandler, CommandExecutionContext executionContext)
Microsoft.VisualStudio.UI.Text.Commanding.Implementation.EditorCommandHandlerService.&lt;&gt;c__DisplayClass14_1`1.&lt;Execute&gt;b__1()
Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call)
--- 
Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Fixes #26388 

The original issue was created by myself. Want to get help for adding tests.

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs [hunk 1]
```diff
@@ -33,8 +33,17 @@ protected override SyntaxNode GetSyntax(SyntaxToken token)
...
+            if (methodDeclaration.Body == null)
+            {
+                return methodDeclaration.GetLocation().SourceSpan.End;
+            }
+            else
+            {
+                // move to the end of the last statement in the method
+                var lastStatement = methodDeclaration.Body.Statements.Last();
+                return lastStatement.GetLocation().SourceSpan.End;
+            }
```
Comment (Neme12) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs]: It looks like this code is same as in OverrideCompletionProvider:
https://github.com/dotnet/roslyn/blob/7fd423dcfacfc0def10f2af07b5bbecab7d51894/src/Features/CSharp/Portable/Completion/CompletionProviders/OverrideCompletionProvider.cs#L196-L206

Consider extracting a helper method to CompletionUtilities.cs such as `GetCaretPositionForDeclaration(SyntaxNode declaration)` by moving the existing `GetTargetCaretPosition` inside `OverrideCompletionProvider` and then calling this helper method from both `OverrideCompletionProvider` and `PartialMethodCompletionProvider`.

Comment (huoyaoyuan) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs]: Yes, I copied from there, and also considered about a helper.

**Comment by huoyaoyuan** [comment]
What's about the ci failure? Seems it's not about my change.

**Comment by Neme12** [comment]
@CyrusNajmabadi @jcouv can you take a look at this?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PartialMethodCompletionProviderTests.cs [hunk 2]
```diff
@@ -475,5 +477,44 @@ partial class PClass
...
+;
+
+                var expected = @"using System;
+partial class Bar
+{
+    partial void Foo();
+    partial void Foo() => throw new NotImplementedException();$$
+}
+"
+;
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PartialMethodCompletionProviderTests.cs]: nit: semicolon should be on the same line

Comment (Neme12) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PartialMethodCompletionProviderTests.cs]: (I know it's in other tests as well, you don't have to fix)

File: src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs]: agreed with @Neme12 .  Please just create a helper for this.

Comment (huoyaoyuan) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/PartialMethodCompletionProvider.cs]: In which class should it be placed? I haven't found a nice place. `CompletionUtilities`?

**Comment by sharwell** [comment]
@jinujoseph for approval

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.preview3

**Pull request merged**